core/sdk-core/src/main/java/software/amazon/awssdk/core/progress/listener/ProgressListener.java

```diff
@@ -0,0 +1,395 @@
+    }
+
+    /**
+    * Additional bytes received
+    * <p>
+    * Available context attributes:
+    * <ol>
+    *     <li>{@link Context.ResponseBytesReceived#request()}</li>
+    *     <li>{@link Context.ResponseBytesReceived#progressSnapshot()}</li>
+    * </ol>
+    */
+    default void attemptFailureResponseBytesReceived(Context.AttemptFailureResponseBytesReceived context) {
+    }
+
+    /**
+     * Successful request execution
+     * <p>
+     * Available context attributes:
+     * <ol>
+     *     <li>{@link Context.ExecutionSuccess#request()}</li>
+     *     <li>{@link Context.ExecutionSuccess#progressSnapshot()}</li>
+     *     <li>{@link Context.ExecutionSuccess#executionSuccess()} ()}</li>
+     * </ol>
+     */
+    default void executionSuccess(Context.ExecutionSuccess context) {
+    }
+
+    /**
+     * This method is called for every failure of a request attempt.
+     * An ideal implementation would invoke executionFailure for a number of attemptFailures greater than a threshold
+     * <p>
+     * Available context attributes:
+     * <ol>
+     *     <li>{@link Context.AttemptFailure#request()}</li>
+     *     <li>{@link Context.AttemptFailure#progressSnapshot()}</li>
+     * </ol>
+     */
+    default void attemptFailure(Context.AttemptFailure context) {
+    }
+
+    /**
+     * A wrapper class that groups together the different context interfaces that are exposed to {@link ProgressListener}s.
+     * <p>
+     * Successful transfer interface hierarchy:
+     * <ol>
+     *     <li>{@link RequestPrepared}</li>
+     *     <li>{@link RequestHeaderSent}</li>
+     *     <li>{@link RequestBytesSent}</li>
+     *     <li>{@link ResponseHeaderReceived}</li>
+     *     <li>{@link ResponseBytesReceived}</li>
+     *     <li>{@link ExecutionSuccess}</li>
+     * </ol>
+     * Failed transfer interface hierarchy:
+     * <ol>
+     *     <li>{@link RequestPrepared}</li>
+     *     <li>{@link AttemptFailure}</li>
+     *     <li>{@link ExecutionFailure}</li>
+     * </ol>
+     *
+     * @see ProgressListener
+     */
+    @SdkProtectedApi
+    final class Context {
+        private Context() {
+        }
+
+        /**
+         * A new transfer has been initiated.
+         * <p>
+         * Available context attributes:
+         * <ol>
+         *     <li>{@link RequestPrepared#request()}</li>
+         *     <li>{@link RequestPrepared#progressSnapshot()}</li>
+         * </ol>
+         */
+        @Immutable
+        @ThreadSafe
+        @SdkPublicApi
+        @SdkPreviewApi
+        public interface RequestPrepared {
+            /**
+             * The {@link SdkHttppRequest} that was submitted to SDK, i.e., the {@link PutObjectRequest} or
+             * {@link GetObjectRequest}.
+             */
+            SdkHttpRequest request();
+
+            /**
+             * The immutable {@link ProgressSnapahot} for this specific update.
+             */
+            ProgressSnapshot progressSnapshot();
```

Comment:
Does this create a new snapshot each time it is called, or is it fixed? (I assume fixed, since it says immutable).

Reply:
We do create a new snapshot for every update. One snapshot corresponds to one state or progress update. For subsequent updates, we create new snapshots. It is immutable once created. Even the updateAndGet method invoked by ProgressUpdater copies the existing snapshot and modifies it.